google_guest_agent/instance_setup.go

```diff
@@ -193,6 +241,307 @@ func agentInit(ctx context.Context) {
+			}
+		}
+	}
+	XPS, err := filepath.Glob(xpsCPU)
+	if err != nil {
+		return err
+	}
+
+	// If we have more CPUs than queues, then stripe CPUs across tx affinity as CPUNumber % queue_count.
+	for _, q := range XPS {
+		numQueues := len(XPS)
+		if numQueues > 63 {
+			numQueues = 63
+		}
+		r, _ := regexp.Compile(queueRegex)
+		match := r.MatchString(q)
+		var queueNum int
+		if match {
+			queueNum, err = strconv.Atoi(r.FindAllStringSubmatch(q, -1)[0][1])
+			if err != nil {
+				return err
+			}
+		}
+		xps := 0
+		for _, cpu := range makeRange(queueNum, numQueues, totalCPUs-1) {
+			xps |= 1 << cpu
+		}
+		// Linux xps_cpus requires a hex number with commas every 32 bits. It ignores
+		// all bits above # cpus, so write a list of comma separated 32 bit hex values
+		// with a comma between dwords.
+		var xpsDwords []string
+		for range makeRange(0, 1, (totalCPUs-1)/32) {
+			xpsDwords = append(xpsDwords, fmt.Sprintf("%08x", xps&0xffffffff))
+		}
+		var xpsString = strings.Join(xpsDwords, ",")
+		if err = ioutil.WriteFile(q, []byte(xpsString), 0644); err != nil {
+			if !strings.Contains(err.Error(), "No such file or directory") {
+				logger.Errorf("Failed to write to xps file, %v", err)
+			}
+			return err
+		}
+		logger.Infof("Queue %d XPS=%s for %s\n", queueNum, xpsString, q)
+	}
+	return nil
+}
+
+func makeRange(min, step, max int) []int {
+	arr := make([]int, (max-min)/step+1)
+	cur := min
+	for i := range arr {
+		arr[i] = cur
+		cur += step
+	}
+	return arr
+}
+
+func copyFile(src string, dest string) error {
+	input, err := ioutil.ReadFile(src)
+	if err != nil {
+		return err
+	}
+	err = ioutil.WriteFile(dest, input, 0644)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func setQueueNumForDevice(dev string) error {
+	dev = path.Base(dev)
+	irqDirs, err := filepath.Glob(irqDirPath)
+	if err != nil {
+		return err
+	}
+	for _, irq := range irqDirs {
+		smpAffinity := irq + "/smp_affinity_list"
+		if isDir(smpAffinity) {
+			continue
+		}
+		virtionetIntxDir := irq + "/" + dev
+		virtionetMsixDirRegex := ".*/" + dev + "-(input|output)\\.([0-9]+)$"
+		if isDir(virtionetIntxDir) {
+			// All virtionet intx IRQs are delivered to CPU 0
+			logger.Infof("Setting %s to 01 for device %s.", smpAffinity, dev)
+			if err := ioutil.WriteFile(smpAffinity, []byte("01"), 0644); err != nil {
+				return err
+			}
+			continue
+		}
+		// Not virtionet intx, probe for MSI-X
+		virtionetMsixFound := 0
+		irqDevs, err := filepath.Glob(irq + "/" + dev + "*")
+		if err != nil {
+			return err
+		}
+		var queueNum int
+		for _, entry := range irqDevs {
+			r, _ := regexp.Compile(virtionetMsixDirRegex)
+			match := r.MatchString(entry)
+			if match {
+				virtionetMsixFound = 1
+				// FindAllStringSubmatch return [][]string
+				queueNum, err = strconv.Atoi(r.FindAllStringSubmatch(entry, -1)[0][2])
+				if err != nil {
+					return err
+				}
+			}
+		}
+		affinityHint := irq + "/affinity_hint"
+		if virtionetMsixFound == 0 || isFile(affinityHint) {
```

Comment:
the shell code `! -f affinity_hint` means "if the file doesn't exist". here you've flipped it